File: cabal/packages/numbers-utils/tests/Operations/FibonacciUtilsTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Operations.FibonacciUtilsTests (fibonacciTest, fibonacciRandomTest) where

import Data.List (isInfixOf, isPrefixOf)

import Data.Time.Clock (getCurrentTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Test.Tasty (TestTree)
import Test.Tasty.HUnit (assertBool, testCase, assertEqual)

import Numbers.Utils.Operations.FibonacciUtils (fibonacci, fibonacciRandom)

formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

fibonacciTest :: TestTree
fibonacciTest = testCase "Fibonacci" $ do
    let n = 100
    let expectedOutput = read "354224848179261915075" :: Integer
    let actualOutput = fibonacci n
    errorMessage <- formatErrorMessage (show expectedOutput) (show actualOutput)
    assertEqual errorMessage expectedOutput actualOutput

fibonacciRandomTest :: TestTree
fibonacciRandomTest = testCase "Fibonacci Random Test" $ do
    result <- fibonacciRandom
    let startString = "Fib["
    let errorMessage = "\nExpected: " ++ startString ++ "\n     Got: " ++ result ++ "\n"
    assertBool errorMessage (startString `isPrefixOf` result)
    assertBool errorMessage (" = " `isInfixOf` result)

File: dotnet/modules/Numbers.Utils/Numbers.Utils.Tests/Operations.Tests/Fibonacci.Utils.Tests/FibonacciUtilsTests.cs
using System.Numerics;

namespace Prog.Numbers.Utils.Operations.Tests;

[TestFixture]
public class FibonacciUtilsTests
{
    private static string FormatErrorMessage(BigInteger expected, BigInteger actual)
    {
        return $"\nTest run at: {DateTime.Now:O}\nExpected : {expected}\nActual   : {actual}\n";
    }

    [Test]
    public void FibonacciTest()
    {
        int n = 100;
        BigInteger expectedOutput = BigInteger.Parse("354224848179261915075");
        BigInteger actualOutput = FibonacciUtils.Fibonacci(n);
        string errorMessage = FormatErrorMessage(expectedOutput, actualOutput);
        Assert.That(expectedOutput, Is.EqualTo(actualOutput), errorMessage);
    }

    [Test]
    public void FinonacciRandomTest()
    {
        string result = FibonacciUtils.FibonacciRandom();
        string startString = "Fib[";
        Assert.That(
            result,
            Does.StartWith(startString),
            $"\nExpected: {startString}\n     Got: {result}\n"
        );
        Assert.That(result, Does.Contain(" = "));
    }
}
